lib/kvs/kvs_cursor.c

```diff
@@ -1100,32 +1100,63 @@ ikvs_cursor_seek(struct kvs_cursor_impl *cursor, const void *key, size_t klen)
     return err;
 }
 
-static merr_t
-ikvs_cursor_kv_copy(struct kvs_cursor_impl *cursor, struct kvs_kvtuple *kvt)
+void
+kvs_cursor_key_copy(struct hse_kvs_cursor *cursor, const void *kbuf, size_t kbufsz, const void **key_out, size_t *klen_out)
 {
-    struct kvs_vtuple *vt = &cursor->kci_elem_last.kce_vt;
-    uint               clen = cursor->kci_elem_last.kce_complen;
-    void *             vbuf = cursor->kci_buf + HSE_KVS_KEY_LEN_MAX;
+    struct kvs_cursor_impl *cur = cursor_h2r(cursor);
+    void                   *buf = (void *)kbuf;
+    size_t                  bufsz = kbufsz;
+    uint                    keylen;
+    const void             *key;
+
+    if (!buf) {
+        buf = (void *)cur->kci_buf;
+        bufsz = HSE_KVS_KEY_LEN_MAX;
+    }
+
+    key = key_obj_copy(buf, bufsz, &keylen, cur->kci_last);
+    *klen_out = keylen;
+
+    if (key_out)
+        *key_out = key;
+}
+
+merr_t
+kvs_cursor_val_copy(struct hse_kvs_cursor *cursor, const void *vbuf, size_t vbufsz, const void **val_out, size_t *vlen_out)
+{
+    struct kvs_cursor_impl *cur = cursor_h2r(cursor);
+
+    struct kvs_vtuple *vt = &cur->kci_elem_last.kce_vt;
+    uint               clen = cur->kci_elem_last.kce_complen;
+    void              *buf = (void *)vbuf;
+    size_t             bufsz = vbufsz;
     merr_t             err = 0;
 
-    kvt->kvt_key.kt_data = key_obj_copy(
-        cursor->kci_buf, HSE_KVS_KEY_LEN_MAX, (uint *)&kvt->kvt_key.kt_len, cursor->kci_last);
+    if (!buf) {
+        buf = cur->kci_buf + HSE_KVS_KEY_LEN_MAX;
+        bufsz = HSE_KVS_VALUE_LEN_MAX;
```

Comment:
`kci_buf` is a flexible array at the end of `struct kvs_cursor_impl` and is always `HSE_KVS_KEY_LEN_MAX + HSE_KVS_VALUE_LEN_MAX` bytes in size.  Since the size is fixed, I don't see any benefit to using a flexible array.  The code would be simplified if `kci_key[HSE_KVS_KEY_LEN_MAX]` and `kci_val[HSE_KVS_VALUE_LEN_MAX ]` were defined in the struct instead of the flexible array.

Reply:
Agree with you